pkg/agent/config.go

```diff
@@ -25,10 +31,76 @@ type Endpoint struct {
 }
 
 type dataGatherer struct {
-	Kind     string `yaml:"kind"`
-	Name     string `yaml:"name"`
-	DataPath string `yaml:"data-path,omitempty"`
-	Config   []byte `yaml:"config"`
+	Kind     string
+	Name     string
+	DataPath string
+	Config   datagatherer.Config
+}
+
+// UnmarshalYAML unmarshals a dataGatherer resolving the type according to Kind.
+func (dg *dataGatherer) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	aux := struct {
+		Kind      string      `yaml:"kind"`
+		Name      string      `yaml:"name"`
+		DataPath  string      `yaml:"data-path,omitempty"`
+		RawConfig interface{} `yaml:"config"`
+	}{}
+	err := unmarshal(&aux)
+	if err != nil {
+		return err
+	}
+
+	dg.Kind = aux.Kind
+	dg.Name = aux.Name
+	dg.DataPath = aux.DataPath
+
+	switch dg.Kind {
+	case "gke":
+		cfg := struct {
+			Config *gke.Config `yaml:"config"`
+		}{}
+		err = unmarshal(&cfg)
+		dg.Config = cfg.Config
+	case "eks":
+		cfg := struct {
+			Config *eks.Config `yaml:"config"`
+		}{}
+		err = unmarshal(&cfg)
+		dg.Config = cfg.Config
+	case "aks":
+		cfg := struct {
+			Config *aks.Config `yaml:"config"`
+		}{}
+		err = unmarshal(&cfg)
+		dg.Config = cfg.Config
+	case "k8s":
+		cfg := struct {
+			Config *k8s.Config `yaml:"config"`
+		}{}
+		err = unmarshal(&cfg)
+		dg.Config = cfg.Config
+	case "local":
+		cfg := struct {
+			Config *local.Config `yaml:"config"`
+		}{}
+		err = unmarshal(&cfg)
+		dg.Config = cfg.Config
+	// dummy dataGatherer is just used for testing
+	case "dummy":
+		cfg := struct {
+			Config *dummyConfig `yaml:"config"`
+		}{}
+		err = unmarshal(&cfg)
+		dg.Config = cfg.Config
```

Comment:
I am sure there is a better way of doing this with `reflect`, but I am too tired at the moment.

I will revisit this tomorrow.

Reply:
I have reworked this part. Hopefully, it is better now.

Reply:
The new version looks :+1: